### ISSUE TITLE
Global API search

### ISSUE THREAD

**Issue by roshanjames** [ocaml/merlin#61]
Consider building a global API search. Merlin has all of the infrastructure required for this. TBD.

**Comment by ELLIOTTCABLE** [comment]
Just wanted to link this to the issue I opened via the VSCode OCaml Platform; "list all global references" would be absolutely amazing:

ocaml/ocaml-lsp#646

**Comment by ELLIOTTCABLE** [comment]
Hey there! This has recently become somewhat critical for refactoring one of our projects.

Happy to contribute back to the ecosystem, if it's within my ablity — @roshanjames pointed out that "Merlin has all of the infrastructure required for this"; is this indeed the case? Any pointers towards getting started, if I wanted to dive in and implement this feature?

**Comment by ulugbekna** [comment]
What a coincidence - https://github.com/ocaml/dune/pull/6762

**Comment by orbitz** [comment]
Does this mean this functionality would require Dune and would not directly be part of Merlin?

**Comment by ELLIOTTCABLE** [comment]
I've been looking into his work; it includes forks of Merlin, ocaml-lsp, and so on to integrate. However, it does require the indexes be generated, and that's indeed dependent on Dune.

**Comment by voodoos** [comment]
> Does this mean this functionality would require Dune and would not directly be part of Merlin?

It requires an indexing phase. Indexation of individual CMT files and merging of indexes is done by a new independent binary. This binary needs to be adequately driven on a given project to generate the index. We are going to provide a set of Dune rules that builds the index, but any other build system could be adapted to do the same.

Everything is still WIP but we plan to provide a simple way to test the current dev version soon!

**Comment by orbitz** [comment]
Great!  Thank you for the hard work.  For various reason, I am using a Makefile-based build so I only ask to keep us non-Dune-folks in mind.  But sounds like you have.  Again, thanks, I'm really looking forward to this feature.

**Comment by ELLIOTTCABLE** [comment]
> Just to be clear, the first version of the index aims at searching all usages of a specific definition which is primarily a navigation tool.

Yep, totally aware — haven't completed evaluation yet, but tentatively, we may allow Dune to orchestrate `ocaml-uideps`, then script consumption of the `.uideps` files.

(We mainly want information about _unreferenced_ values; which, based on my understanding, _is_ something that we can extract out of your work. At least, it's sure looking like it thus far! If this works out well, would you likely accept a PR providing something like `ocaml-uideps --unused` to dump a report on interface-items with no external references?)

**Comment by voodoos** [comment]
> (We mainly want information about _unreferenced_ values; which, based on my understanding, _is_ something that we can extract out of your work. At least, it's sure looking like it thus far! If this works out well, would you likely accept a PR providing something like `ocaml-uideps --unused` to dump a report on interface-items with no external references?)

Yes, the index could eventually be used to list values that are exported by a module but never referenced by others.

**Comment by NoahTheDuke** [comment]
Now that https://github.com/ocaml/dune/pull/10422 has been merged, what's the next step for support in Merlin? Is this something that anyone can contribute to?

**Comment by voodoos** [comment]
Well it's already available 🙂

Announcement: https://discuss.ocaml.org/t/ann-project-wide-occurrences-in-merlin-and-lsp/14847

It integrates naturally with LSP's reference queries but we also provide new commands in the custom emacs and vim plugins:
https://ocaml.github.io/merlin/editor/emacs/#search-for-an-identifiers-occurrences
https://ocaml.github.io/merlin/editor/vim/#search-for-an-identifiers-occurrences

In any case, for the feature to work, it is necessary to have an up-to-date index, I recommend running `dune build @ocaml-index --watch` while developing for projects that don't cost too much to index.

Finally, since the latest release, Merlin also has experimental support for **renaming**:
https://discuss.ocaml.org/t/ann-merlin-and-ocaml-lsp-support-experimental-project-wide-renaming/16008

Please experiment with these new exciting features and give your feedback in the pinned issue: 
https://github.com/ocaml/merlin/issues/1780

I am also interested to know what we could have done for you to more easily discover that this feature actually exists :-)

**Comment by orbitz** [comment]
Wow great to see this!

Sorry if this not the correct place to ask this but are there directions for how to use this without dune?  I use ocamllsp with --fallback-to-dot-merlin enabled.  Can I just run the indexer in my directory with .merlin and magic happens?

Thank you for the great work!

**Comment by NoahTheDuke** [comment]
Fantastic news, @voodoos! Thanks for the links, those are helpful. I came across this issue through https://github.com/ocaml/ocaml-lsp/issues/646, which is the way that I interact with Merlin. Given that this issue and the ocaml-lsp issues are still open, I didn't do any further research and assumed that they were still in development.

I'll check in with the ocaml-lsp folks to see about using the new features!
